Prepare Dataset View

MillarayVVS opened the pull request:
# Summary

<!--
Include a short summary of the changes made to the platforn and list any dependencies
of the pr - Required.
-->
Adds Prepare dataset step in the New Experiment flow

## Type of change

<!-- Indicate the type of change. Delete those that do not apply  - Required -->

- Front end new feature.

## Changes

<!--
Indicate the changes/fixes that you want to merge - Required.

- Bug 1
- Bug 2
- Feature 1
- Feature 2
-->
- Insert dataset columns that you wanna use as input for the model.
- Insert dataset columns that you wanna use as output or class label for the model.
- Insert splits using random rows or specific for each split between training, validation and testing.
## How to Test

<!--
Describe the tests or executions that you ran to verify your changes and provide
instructions to reproduce them - Required.
-->
1. Go to ```DashAI/DashAI/front``` and run ```yarn build```
2. Go back to ```DashAI/``` and run ```python -c "import DashAI;DashAI.run()"```
3. Check the view in New Experiment.

## Screenshots

<!--
In the case of modifying a view in the front-end, include screenshots with the changes.
Optional, delete this section if not needed.-->
![image](https://github.com/DashAISoftware/DashAI/assets/70553082/3352b04e-9834-4120-9cd4-f7c9ab7646e4)
## Notes

<!--
Include any additional information that would be useful to the reviewer.
Optional, delete this section if not needed.
-->
- There's still things that needs to be added to this new feature such as validate the rows types with the task, evade overlaping between column and rows splits, etc.

MillarayVVS committed:
Remove Select Task step from New Dataset flow

MillarayVVS committed:
Remove Select Task step from New Dataset flow

MillarayVVS committed:
List all datasets in Step 2 of New Experiment

MillarayVVS committed:
List all datasets in Step 2 of New Experiment

MillarayVVS committed:
Adds Prepare Dataset step to New Experiment flow

MillarayVVS committed:
Adds Prepare Dataset step to New Experiment flow

Commit:
CLI

Commit:
Merge branch 'staging' into feature/cli

MillarayVVS committed:
Add parse to input and output column range and numbers

MillarayVVS committed:
Fix merge

MillarayVVS committed:
Add logic and parse to dataset splits

cristian-tamblay committed:
Changed readme to reflect these changes

cristian-tamblay committed:
Merge pull request #154 from DashAISoftware/feature/cli

Command Line Interface for DashAI

MillarayVVS committed:
Adding change and error handling to split rows

MillarayVVS committed:
Merge branch 'back/modify-db' into front/dataset-new-flow

MillarayVVS committed:
Complete de Prepare dataset step with the input and output columns values and split

MillarayVVS committed:
Fix New Experiment load with the new params

pbadillatorrealba commented on DashAI/front/src/components/experiments/ConfigureModelsStep.jsx:
```diff
@@ -157,6 +157,9 @@ ConfigureModelsStep.propTypes = {
     dataset: PropTypes.object,
     task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
```
Prefer to use number from proptypes.

```suggestion
    input_columns: PropTypes.arrayOf(PropTypes.number),
    output_columns: PropTypes.arrayOf(PropTypes.number),
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/ModelsTable.jsx:
```diff
@@ -109,6 +109,9 @@ ModelsTable.propTypes = {
     dataset: PropTypes.object,
     task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
```
```suggestion
    input_columns: PropTypes.arrayOf(PropTypes.number),
    output_columns: PropTypes.arrayOf(PropTypes.number),
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/ModelsTable.jsx:
```diff
@@ -109,6 +109,9 @@ ModelsTable.propTypes = {
     task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
+    splits: PropTypes.object,
```
Maybe you could define here that split is a json with train,validation and test.

```suggestion
    splits: PropTypes.shape({
        train: PropTypes.number,
        validation: PropTypes.number,
        test: PropTypes.number
  })
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/NewExperimentModal.jsx:
```diff
@@ -94,6 +99,9 @@ export default function NewExperimentModal({
         newExp.dataset.id,
         newExp.task_name,
         newExp.name,
+        JSON.stringify(newExp.input_columns),
+        JSON.stringify(newExp.output_columns),
+        JSON.stringify(newExp.splits),
```
Why stringify is required here?

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+    const numbersArray = [];
+    if (!rangeRegex.test(value)) {
+      throw new Error(
+        "Ranges and indexes must match the format on the example above",
```
Maybe put here that that the entered text don't fit the example format, not to mention "the above".

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
This component is extremely complex. Could it be divided into two stages?
Maybe in select columns and then the holdout. 

Or if you want to leave them together, separate the logic of the first and second parts into separate components (files) and then orchestrate them in a step.

MillarayVVS replied:
I divided the Prepare Dataset Step in two different components in the recent commit, can you review it again please.

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+  const handleInputColumnsChange = (event) => {
+    const input = event.target.value.replace(/ /g, ""); // TODO: dont accept spaces between numbers
+    try {
+      const columnNumbers = parseRangeToIndex(input, totalColumns);
```
`columnIndex` (same as the next function)?

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+    }
+  };
+  const handleRowsPreferenceChange = (event) => {
+    if (event.target.value === "manually") {
```
I think this would look better as `splitByIndex`.
```suggestion
    if (event.target.value === "splitByIndex") {
```

Therefore, the other option could be called `splitByPercentage.`

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+            ? { ...rowsPartitionsIndex, training: rowsIndex }
+            : id === "validation"
+            ? { ...rowsPartitionsIndex, validation: rowsIndex }
+            : { ...rowsPartitionsIndex, testing: rowsIndex },
```
It's hard to read code like this! 
I don't think it's good to use 2 ternary operators in a row. 
I think there are better strategies like a switch case or something similar.


```suggestion
switch (id) {
  case "training":
    setRowsPartitionsIndex({ ...rowsPartitionsIndex, training: rowsIndex });
    break;
  case "validation":
    setRowsPartitionsIndex({ ...rowsPartitionsIndex, validation: rowsIndex });
    break;
  case "test":
    setRowsPartitionsIndex({ ...rowsPartitionsIndex, testing: rowsIndex });
    break;
}
```

This also applies to the following functions.

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+      <Grid container spacing={1}>
+        <Grid item xs={12}>
+          <Typography item variant="subtitle1" component="h3" sx={{ mb: 0 }}>
+            Choose which columns to use
```
```suggestion
            Indicate which columns of the dataset will be used as input and output.
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+            component="h3"
+            sx={{ mb: 2, color: "grey" }}
+          >
+            Ex: 1-6, 23-108
```
```suggestion
            The notation is based on ranges. For example: 1-6, 23-108
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+        <Grid container spacing={1}>
+          <Grid item xs={12}>
+            <Typography variant="subtitle1" component="h3" sx={{ mb: 2 }}>
+              Choose how you wanna divide the dataset for the experiment
```
```suggestion
              Select how to divide the dataset into training, validation and test subsets.
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+          />
+          {rowsPreference === "random" ? (
+            <React.Fragment>
+              <Grid container direction="row" spacing={4}>
```
The options look very close together. How about adding some small margins?

```suggestion
              <Grid container direction="row" spacing={4} sx={{my: 2}}>
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+            <React.Fragment>
+              <Grid container direction="row" spacing={4}>
+                <Grid item sx={{ xs: 4 }}>
+                  <TextField
+                    id="training"
+                    label="Training"
+                    autoComplete="off"
+                    error={rowsPartitionsError}
+                    onChange={handleRowsChange}
+                  />
```
Some input comments: 

- You can use certain types to render a specific component for the selected type, for example number (type param).
- I think that perhaps compact inputs would look better (size param).
- You could add start and end adornments by using InputProps.

```suggestion
                  <TextField
                    id="training"
                    label="Training"
                    autoComplete="off"
                    type="number"
                    size="small"
                    InputProps={{
                      startAdornment: <InputAdornment position="end">%</InputAdornment>,
                    }}
                    error={rowsPartitionsError}
                    onChange={handleRowsChange}
                  />
```

If you change it, remember to change them all!

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+          </Grid>
+        </Grid>
+        <RadioGroup
+          aria-labelledby="demo-radio-buttons-group-label"
```
```suggestion
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+          <FormControlLabel
+            value="manually"
+            control={<Radio />}
+            label="Introduce splits manually by rows index"
```
```suggestion
            label="Use manual splitting by specifying the row indexes of each subset"
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+    name: PropTypes.string,
+    dataset: PropTypes.object,
+    task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
```
```suggestion
    input_columns: PropTypes.arrayOf(Proptypes.number),
    output_columns: PropTypes.arrayOf(Proptypes.number),
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+    task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
+    splits: PropTypes.object,
```
Check the past proptypes for split.

pbadillatorrealba commented on DashAI/front/src/components/experiments/SelectDatasetStep.jsx:
```diff
@@ -173,6 +166,9 @@ SelectDatasetStep.propTypes = {
     name: PropTypes.string,
     dataset: PropTypes.object,
     task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
+    splits: PropTypes.object,
```
Same here with types.

pbadillatorrealba commented on DashAI/front/src/components/experiments/SetNameAndTaskStep.jsx:
```diff
@@ -150,6 +150,9 @@ SetNameAndTaskStep.propTypes = {
     name: PropTypes.string,
     dataset: PropTypes.object,
     task_name: PropTypes.string,
+    input_columns: PropTypes.arrayOf(number),
+    output_columns: PropTypes.arrayOf(number),
+    splits: PropTypes.object,
```
Same here.

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+  };
+
+  const checkSplit = (training, validation, testing) => {
+    if (parseInt(training) + parseInt(validation) + parseInt(testing) === 100) {
+      return true;
+    } else {
+      return false;
+    }
```
It can be shortened a little without losing readability.
```suggestion
    return parseInt(training) + parseInt(validation) + parseInt(testing) === 100
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+    validation: [],
+    testing: [],
+  };
+  const defaultPartitionsPercentage = {
+    training: 0,
+    validation: 0,
+    testing: 0,
+  };
```
Why not use the standard ones right away? For example: 60, 20, 20?

```suggestion
  const defaultPartitionsPercentage = {
    training: 60,
    validation: 20,
    testing: 20,
  };
```

pbadillatorrealba commented on DashAI/front/src/components/experiments/PrepareDatasetStep.jsx:
```diff
@@ -0,0 +1,398 @@
+ */
+function PrepareDatasetStep({ newExp, setNewExp, setNextEnabled }) {
+  // TODO: column and row numbers should be minor to the maximum on the dataset
+  const totalColumns = 100;
+  const totalRows = 2000;
```
What is the meaning of these variables?

pbadillatorrealba reviewed (changes requested):
Implementation looks good :)

I left a lot of comments to talk about.

nataliabarca committed:
Back/predict endpoint (#121)

* Predict endpoint for tabular classification

* Predict endpoint for tabular classification tasks

* Fixed parameters in predict POST request

* Changed dataset name in predict test

* fix import to pass pre-commit

* delete duplicated test, the runner migth no be tested in the predict tests

* add new test dataset to use in predict endpoint

* add PredictParams pydantic model to use in predict POST endpoint

* remove deprecated test for runner endpoint

* change dataset, fix trained_run_id to execture the job and change predict test to fit new format

* get features_names from run dataset, load input file into DashAIDataset, reorder columns to fit the dataset used in train and test and return a list with the prediction of each input example

* delete asserts to check predict return labels, TBD in another PR

* add documentation, not implemented CRUD endpoints and tests

---------

Co-authored-by: natalia.abarca <natalia.abarca@zippedi.com>
Co-authored-by: Lurrea <rodrigo.urrea@ug.uchile.cl>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/ConfigureModelsStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/ModelsTable.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/ModelsTable.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Update DashAI/front/src/components/experiments/PrepareDatasetStep.jsx

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Fix code changes

MillarayVVS committed:
Divide Prepare Dataset Step in two components

MillarayVVS committed:
Fix Dataset and Experiments types

MillarayVVS committed:
Adding styling components to Splits inputs

pbadillatorrealba commented:
@MillarayVVS , I can check your changes again, but first merge the changes that are staging.

Lucas-CE committed:
BaseConverter and DropColumns (#153)

* feat: :sparkles: Preprocessor class creation v1

* revert: :rewind: Revert preprocessor class commit

* feat: :sparkles: BasePreprocessor class creation

* feat: :sparkles: Remove Column Preprocessor creation

The preprocessor operates according to Millaray mockups, employing the initial and concluding indices for column removal. This allows us to employ various remove preprocessors to eliminate specific sets of columns, such as removing columns 1-10 and 49-80, for instance.

* BaseProcessor extends ConfigObject and ABCMeta

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* Update transform function name and signature

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* fix: :bug: Fixing DashAIDataset not defined error

* refactor: :recycle: Preprocess names are changed to transform

* refactor: :truck: Renaming files to transformers

* feat: :sparkles: In the drop column transformer the transform signature is fixed, and allows other input types

* fix: :adhesive_bandage: Fix precommit error of multiple isinstance uses

* fix the transform signature in DropColumnTransformer

* Dropping columns process is separated depending on whether they do it with name or if they do it with index

* fit method implemented in transformers

* transformers now use beartype notation

* fix: :bug: Dashai imports must be done from the DashAI root to avoid problems with the tests

* refactor: :adhesive_bandage: Correcting methods input types

* fix: :bug: Droppers updated to remove the column name from the input columns list

* refactor: :recycle: The transform function was duplicated

* refactor: :recycle: To remove input columns must use a columns list

* test: :white_check_mark: Drop column transformer first test implemented

* refactor: :recycle: The datasetdicts that enter to transformers dont have input and output columns

* refactor: :truck: The transformers are renamed to Converters

* Fixing base_converter annotation

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* Change parameter name in ColumnDropperByIndex init

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* Use named parameters in converter tests

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* Change parameter name in ColumnDropperByName init

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* Fix type annotation fit function in ColumnDropperByIndex

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

* fix: :bug: Fixed type errors and parameter names errors

* refactor: :recycle: Init creation to preprocessor folder

* feat: :sparkles: ColumnDropperByIndex can receive one column index

* test: :white_check_mark: Adding ColumnDropperByIndex test and changing fixture scope in test

* test: :white_check_mark: Test to drop column by index range

* test: :white_check_mark: Test to drop columns converter using list of names

* refactor: :recycle: Renaming transformer and preprocessor folders and files

* docs: :memo: Fixing little documentation errors in ColumnDropperByIndex constructor

* refactor: :fire: Remove .arrow files generated in testing

---------

Co-authored-by: Pablo Badilla Torrealba <pbadilla.torrealba@gmail.com>

MillarayVVS committed:
Merge branch 'staging' into front/dataset-new-flow

Riul1999 commented on DashAI/back/api/api_v1/schemas/experiments_params.py:
Why are columns and splits str?

MillarayVVS committed:
Fix Pytests

MillarayVVS committed:
Adds new method get_dataset_info

MillarayVVS committed:
Get the total of columns and rows for validation

MillarayVVS committed:
Fix Input and Output default values

MillarayVVS committed:
Change input, output and split type to JSON

MillarayVVS committed:
Merge branch 'project/datasets_rework' into front/dataset-new-flow

MillarayVVS committed:
Fix precommit format
